### ISSUE TITLE
package or namespace load failed (GOSemSim and RxnSim)

### ISSUE THREAD

**Issue by ksrinivs64** [opencpu/opencpu#177]
Hello, I followed instructions to install GOSemSim (http://bioconductor.org/packages/release/bioc/html/GOSemSim.html) and RxnSim (using the usual install.packages route).  I am installing as root.  In my interactive R session library(GOSemSim) and library(RxnSim) work but when I try to access it using opencpu, I get:
package or namespace load failed for 'RxnSim'

I do have the directories: /usr/local/lib/R/site-library/RxnSim/ and /usr/local/lib/R/site-library/GOSemSim/, and I have restarted opencpu and apache2 several times to no avail.  One thing I noticed is the bioclite had a warning that opencpu was newer and asked me to downgrade the package.  Could that be an issue?  Thanks!

**Comment by jeroenooms** [comment]
> One thing I noticed is the bioclite had a warning that opencpu was newer and asked me to downgrade the package. Could that be an issue? 

No that should not be an issue. Can you check if there are any warnings in any of the log files? In particular 

```
/var/log/kern.log
/var/log/apache2/error.log
/var/log/opencpu/error.log
```

**Comment by ksrinivs64** [comment]
Mar  7 19:47:26 mac1 kernel: [5285357.003033] type=1400 audit(1457401646.084:133): apparmor="DENIED" operation="file_lock" profile="opencpu-exec" name="/usr/local/lib/R/site-library/GO.db/extdata/GO.sqlite" pid=18182 comm="apache2" requested_mask="k" denied_mask="k" fsuid=33 ouid=0
Seems to be related to at least the GOSemSim failure.  This appeared in /var/log/kern.log.  Opencpu/error.log has nothing in it, and the apache2/error.log didn't have anything suspicious.

**Comment by jeroenooms** [comment]
OK this looks like a security problem. Have a look at section 3.5 of the [pdf manual](http://jeroenooms.github.com/opencpu-manual/opencpu-server.pdf). Also [this post](https://groups.google.com/forum/#!topic/opencpu/SllsD379Zdc) about sqlite perhaps. I think you need to edit `/etc/apparmor.d/opencpu.d/custom` and add a rule like this:

```
/usr/local/lib/R/site-library/** rmk
```

**Comment by ksrinivs64** [comment]
Thanks, that fixed the issue with accessing GOSemSim.  I still have the failure with RxnSim with no errors in any of the logs.  Based on what I saw here about rJava, and knowing that RxnSim has a dependency ultimately on rJava, I followed your advice and tried to call the function directly using: RxnSim::ms.compute.  Now, I can see the error: 
.onLoad failed in loadNamespace() for 'rJava', details:
  call: dyn.load(file, DLLpath = DLLpath, ...)
  error: unable to load shared object '/usr/local/lib/R/site-library/rJava/libs/rJava.so':
  libjvm.so: cannot open shared object file: No such file or directory
I tried setting the LD_LIBRARY_PATH, called sudo R CMD javareconf, and even sys linked the file into the R libraries as one of the users suggested.  I've restarted opencpu since, but it still fails in opencpu but not in the R command line.  Is there some configuration in opencpu that I should be setting?  Thanks for your help!

**Comment by jeroenooms** [comment]
You're sure there are no additional security errors in `kern.log`? Where on your disk is `libjvm.so` located?

**Comment by ksrinivs64** [comment]
Its in /usr/lib/jvm/java-ibm-1.7/jre/lib/amd64/j9vm/libjvm.so.  
echo $LD_LIBRARY_PATH
/usr/lib/jvm/java-ibm-1.7/jre/lib/amd64/j9vm/
Same error as earlier:  error: unable to load shared object '/usr/local/lib/R/site-library/rJava/libs/rJava.so':
  libjvm.so: cannot open shared object file: No such file or directory
But nothing in kern.log other than: 
Mar  8 09:24:06 rdfstore kernel: [5334357.793550] type=1400 audit(1457450646.876:187): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="/usr/sbin/cupsd" pid=12682 comm="apparmor_parser"
Current time on that machine is 9:57.

**Comment by jeroenooms** [comment]
Is there a reason you're not using openjdk from ubuntu? That one should just install to the global `/usr/lib`. Anyway maybe you can symlink it, i.e:

```
sudo ln -s /usr/lib/jvm/java-ibm-1.7/jre/lib/amd64/j9vm/libjvm.so /usr/local/lib/libjvm.so 
```

**Comment by jeroenooms** [comment]
You can also try to set `LD_LIBRARY_PATH` in `/etc/opencpu/Renviron`...

**Comment by jeroenooms** [comment]
But if you can make it work without Java I highly recommend that :-)

**Comment by ksrinivs64** [comment]
The ld library is working now but unfortunately I get: R call failed: process died.  I did try to change rlimit.as to 3e9 but that didnt make any difference.  Nothing in kern.log, like others have reported.  Just out of curiosity, why does this package work on the R command line but not in opencpu?

**Comment by ksrinivs64** [comment]
I should have added that I hardly want to use Java :) but I guess this package that does some chemical fingerprint algorithm does.  And there seems to be no pure R alternative.

**Comment by jeroenooms** [comment]
The OpenCPU server forks the main process, however the JVM cannot be forked. See also https://github.com/s-u/rJava/issues/25

**Comment by ksrinivs64** [comment]
I see, so are we out of luck for running this library through opencpu?

**Comment by jeroenooms** [comment]
Well the one thing you can try is lazy load the library so that the jvm doesn't started until after the fork. So that means not importing it in your `NAMESPACE` but instead calling RxnSim::fun() in your package...

**Comment by ksrinivs64** [comment]
I already did that.  Same error despite that.

**Comment by jeroenooms** [comment]
Hmm yes I think you're out of luck. Sorry about that... Java is always a pain in R...

**Comment by ksrinivs64** [comment]
No problem - thanks for your help!  Great library by the way.

**Comment by jeroen** [comment]
The Java problem should be solved in OpenCPU 2.0
